File: WireExtensionComponents/Views/UserImageView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 

@import WireSyncEngine;

#import "UserImageView.h"
#import "UIImage+ImageUtilities.h"
#import "UIImage+ZetaIconsNeue.h"

#import "weakify.h"

#import <WireExtensionComponents/WireExtensionComponents-Swift.h>

@interface UserImageView ()

@property (nonatomic) id userObserverToken;
@property (nonatomic) RoundedView *indicator;

@end



@implementation UserImageView

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self setupBasicProperties];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupBasicProperties];
    }
    return self;
}

- (instancetype)initWithSize:(UserImageViewSize)size
{
    self = [self initWithFrame:CGRectZero];

    if (self) {
        self.size = size;
    }

    return self;
}

- (void)setupBasicProperties
{
    _shouldDesaturate = YES;
    _size = UserImageViewSizeSmall;

    self.accessibilityElementsHidden = YES;
    
    [self createIndicator];
    [self configureConstraints];
}

- (CGSize)intrinsicContentSize
{
    CGFloat imageSize = self.size;
    return CGSizeMake(imageSize, imageSize);
}

- (void)setUser:(id<UserType, AccentColorProvider>)user
{    
    _user = user;
    
    if (self.userSession != nil && user != nil && ([user isKindOfClass:[ZMUser class]] || [user isKindOfClass:[ZMSearchUser class]])) {
        self.userObserverToken = [UserChangeInfo addObserver:self forUser:user userSession:self.userSession];
    }
    
    self.initials.textColor = UIColor.whiteColor;
    self.initials.text = user.initials.uppercaseString;
    
    self.imageView.contentMode = UIViewContentModeScaleAspectFill;
    [self updateForServiceUserIfNeeded:user];
    [self setUserImage:nil animated:NO];
    [self updateIndicatorColor];
    [self updateUserImage];
}

- (void)updateForServiceUserIfNeeded:(id <UserType>)user
{
    self.shape = [self shapeForUser:user];
    self.containerView.layer.borderColor = [self borderColorForUser:user];
    self.containerView.layer.borderWidth = [self borderWidthForUser:user];
    self.imageView.backgroundColor = [self containerBackgroundColorForUser:user];
}

- (AvatarImageViewShape)shapeForUser:(id <UserType>)user
{
    return user.isServiceUser ? AvatarImageViewShapeRoundedRelative : AvatarImageViewShapeCircle;
}

- (UIColor *)containerBackgroundColorForUser:(id <UserType>)user
{
    return user.isServiceUser ? UIColor.whiteColor : UIColor.clearColor;
}

- (CGColorRef)borderColorForUser:(id <UserType>)user
{
    return user.isServiceUser ? [UIColor.blackColor colorWithAlphaComponent:0.08].CGColor : nil;
}

- (CGFloat)borderWidthForUser:(id <UserType>)user
{
    return user.isServiceUser ? 0.5 : 0;
}

- (void)createIndicator
{
    self.indicator = [RoundedView new];
    self.indicator.backgroundColor = UIColor.redColor;
    self.indicator.hidden = YES;
    [self.indicator toggleCircle];
    
    [self addSubview:self.indicator];
}

- (void)configureConstraints
{
    [self setContentHuggingPriority:1000 forAxis:UILayoutConstraintAxisVertical];
    [self setContentHuggingPriority:1000 forAxis:UILayoutConstraintAxisHorizontal];

    self.indicator.translatesAutoresizingMaskIntoConstraints = NO;

    NSArray<NSLayoutConstraint *> *constraints =
  @[
    [self.indicator.topAnchor constraintEqualToAnchor:self.topAnchor],
    [self.indicator.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
    [self.indicator.widthAnchor constraintEqualToAnchor:self.containerView.widthAnchor multiplier:1/3],
    [self.indicator.heightAnchor constraintEqualToAnchor:self.indicator.heightAnchor]
    ];

    [NSLayoutConstraint activateConstraints:constraints];
}

- (void)updateIndicatorColor
{
    self.indicator.backgroundColor = [(id)self.user accentColor];
}

- (void)setUserImage:(UIImage *)userImage animated:(BOOL)animated
{
    dispatch_block_t imageUpdate = ^{
        self.initials.hidden = userImage != nil;
        self.imageView.hidden = userImage == nil;
        self.imageView.image = userImage;
        
        BOOL isWireless = NO;
        if ([self.user respondsToSelector:@selector(isWirelessUser)]) {
            isWireless = [(id)self.user isWirelessUser];
        }
        
        if (userImage) {
            self.containerView.backgroundColor = [self containerBackgroundColorForUser:self.user];
        }
        else if ([self.user respondsToSelector:@selector(accentColor)] &&
                 (self.user.isConnected || self.user.isSelfUser || self.user.isTeamMember || isWireless)) {
            self.containerView.backgroundColor = [(id)self.user accentColor];
        }
        else {
            self.containerView.backgroundColor = [UIColor colorWithWhite:0.8 alpha:1];
        }
    };
    
    if (animated) {
        [UIView transitionWithView:self duration:0.15 options:UIViewAnimationOptionTransitionCrossDissolve animations:imageUpdate completion:NULL];
    } else {
        imageUpdate();
    }
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    [super touchesEnded:touches withEvent:event];
    
    if ([self.delegate respondsToSelector:@selector(userImageViewTouchUpInside:)]) {
        [self.delegate userImageViewTouchUpInside:self];
    }
}

#pragma mark - Indicator

- (void)setIndicatorEnabled:(BOOL)indicatorEnabled
{
    _indicatorEnabled = indicatorEnabled;
    
    self.indicator.hidden = ! indicatorEnabled;
}

#pragma mark - ZMUserObserver

- (void)userDidChange:(UserChangeInfo *)change
{
    if (self.size == UserImageViewSizeBig) {
        if (change.imageMediumDataChanged || change.connectionStateChanged) {
            [self updateUserImage];
        }
    }
    else {
        if (change.imageSmallProfileDataChanged || change.connectionStateChanged || change.teamsChanged) {
            [self updateUserImage];
        }
    }
    
    if (change.accentColorValueChanged) {
        [self updateIndicatorColor];
    }
}

@end

File: Wire-iOS/Sources/UserInterface/Components/UserImageView.swift
//
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import WireSyncEngine

/**
 * A view that displays the avatar for a remote user.
 */

@objc open class UserImageView: AvatarImageView, ZMUserObserver {

    /**
     * The different sizes for the avatar image.
     */

    @objc(UserImageViewSize) public enum Size: Int {
        case tiny = 16
        case badge = 24
        case small = 32
        case normal = 64
        case big = 320
    }

    // MARK: - Interface Properties

    /// The size of the avatar.
    @objc public var size: Size {
        didSet {
            updateUserImage()
        }
    }

    /// Whether the image should be desaturated, e.g. for unconnected users.
    @objc public var shouldDesaturate: Bool = true

    /// Whether the badge indicator is enabled.
    public var indicatorEnabled: Bool = false {
        didSet {
            badgeIndicator.isHidden = !indicatorEnabled
        }
    }

    private let badgeIndicator = RoundedView()

    // MARK: - Remote User

    /// The user session to use to download images.
    @objc public var userSession: ZMUserSession? {
        didSet {
            updateUser()
        }
    }

    /// The user to display the avatar of.
    @objc public var user: UserType? {
        didSet {
            updateUser()
        }
    }

    private var userObserverToken: Any?

    // MARK: - Initialization

    public override init(frame: CGRect) {
        self.size = .small
        super.init(frame: .zero)
        configureSubviews()
        configureConstraints()
    }

    public init(size: Size = .small) {
        self.size = size
        super.init(frame: .zero)
        configureSubviews()
        configureConstraints()
    }

    deinit {
        userObserverToken = nil
    }

    public required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    open override var intrinsicContentSize: CGSize {
        return CGSize(width: size.rawValue, height: size.rawValue)
    }

    private func configureSubviews() {
        accessibilityElementsHidden = true

        badgeIndicator.backgroundColor = .red
        badgeIndicator.isHidden = true
        badgeIndicator.shape = .circle
        addSubview(badgeIndicator)
    }

    private func configureConstraints() {
        badgeIndicator.translatesAutoresizingMaskIntoConstraints = false

        setContentHuggingPriority(.required, for: .vertical)
        setContentHuggingPriority(.required, for: .horizontal)

        NSLayoutConstraint.activate([
            badgeIndicator.topAnchor.constraint(equalTo: topAnchor),
            badgeIndicator.trailingAnchor.constraint(equalTo: trailingAnchor),
            badgeIndicator.heightAnchor.constraint(equalTo: badgeIndicator.widthAnchor),
            badgeIndicator.widthAnchor.constraint(equalTo: widthAnchor, multiplier: 1/3)
        ])
    }

    // MARK: - Interface

    /// Returns the appropriate border width for the user.
    private func borderWidth(for user: UserType) -> CGFloat {
        return user.isServiceUser ? 0.5 : 0
    }

    /// Returns the appropriate border color for the user.
    private func borderColor(for user: UserType) -> CGColor? {
        return user.isServiceUser ? UIColor.black.withAlphaComponent(0.08).cgColor : nil
    }

    /// Returns the placeholder background color for the user.
    private func containerBackgroundColor(for user: UserType) -> UIColor? {
        let isWireless = user.zmUser?.isWirelessUser == true

        switch self.avatar {
        case .image?, nil:
            return user.isServiceUser ? .white : .clear
        case .text?:
            if user.isConnected || user.isSelfUser || user.isTeamMember || isWireless {
                return user.indexedAccentColor
            } else {
                return UIColor(white: 0.8, alpha: 1)
            }
        }
    }

    /// Returns the appropriate avatar shape for the user.
    private func shape(for user: UserType) -> AvatarImageView.Shape {
        return user.isServiceUser ? .relative : .circle
    }

    // MARK: - Changing the Content

    /**
     * Sets the avatar for the user with an optional animation.
     * - parameter avatar: The avatar of the user.
     * - parameter user: The currently displayed user.
     * - parameter animated: Whether to animate the change.
     */

    func setAvatar(_ avatar: Avatar, user: UserType, animated: Bool) {
        let updateBlock = {
            self.avatar = avatar
            self.container.backgroundColor = self.containerBackgroundColor(for: user)
        }

        if animated {
            UIView.transition(with: self, duration: 0.15, options: .transitionCrossDissolve, animations: updateBlock, completion: nil)
        } else {
            updateBlock()
        }
    }

    /// Updates the image for the user.
    fileprivate func updateUserImage() {
        guard let user = user, let userSession = userSession else { return }

        var desaturate = false
        if shouldDesaturate {
            desaturate = !user.isConnected && !user.isSelfUser && !user.isTeamMember && !user.isServiceUser
        }

        user.fetchProfileImage(session: userSession, sizeLimit: size.rawValue, desaturate: desaturate, completion: { [weak self] (image, cacheHit) in
            // Don't set image if nil or if user has changed during fetch
            guard let image = image, user.isEqual(self?.user) else { return }
            self?.setAvatar(.image(image), user: user, animated: !cacheHit)
        })
    }

    // MARK: - Updates

    open func userDidChange(_ changeInfo: UserChangeInfo) {
        // Check for potential image changes
        if size == .big{
            if changeInfo.imageMediumDataChanged || changeInfo.connectionStateChanged {
                updateUserImage()
            }
        } else {
            if changeInfo.imageSmallProfileDataChanged || changeInfo.connectionStateChanged || changeInfo.teamsChanged {
                updateUserImage()
            }
        }

        // Change for accent color changes
        if changeInfo.accentColorValueChanged {
            updateIndicatorColor()
        }
    }

    /// Called when the user or user session changes.
    open func updateUser() {
        guard let user = self.user, let initials = user.initials else {
            return
        }

        let defaultAvatar = Avatar.text(initials.localizedUppercase)
        setAvatar(defaultAvatar, user: user, animated: false)

        if let userSession = self.userSession {
            userObserverToken = UserChangeInfo.add(observer: self, for: user, userSession: userSession)
        }

        updateForServiceUserIfNeeded(user)
        updateIndicatorColor()
        updateUserImage()
    }

    /// Updates the color of the badge indicator.
    private func updateIndicatorColor() {
        self.badgeIndicator.backgroundColor = user?.indexedAccentColor
    }

    /// Updates the interface to reflect if the user is a service user or not.
    private func updateForServiceUserIfNeeded(_ user: UserType) {
        shape = shape(for: user)
        container.layer.borderColor = borderColor(for: user)
        container.layer.borderWidth = borderWidth(for: user)
        container.backgroundColor = containerBackgroundColor(for: user)
    }

}
